### ISSUE TITLE
Implement C, P, A, and O flags on x86

### ISSUE THREAD

**Issue by occamsrzor** [Vector35/binaryninja-api#4044]

**Comment by fuzyll** [comment]
@rssor is working on part of this on a separate branch for the upcoming release (specific carry flag stuff). I've assigned it to him and stuck it in the milestone so we track that progress, but we'll be removing it from the milestone and un-assigning it once he's completed that work.

**Comment by fuzyll** [comment]
@romanholidaypancakes This is where @rssor made the changes that I had mentioned above: https://github.com/Vector35/arch-x86/commit/5b134ac7da45c009094faf654f613a9ecd11cae8.

If anyone else wants to take a look at this and PR additional changes, we'd really appreciate it.

The parity flag stuff is _probably_ not doable without some far more invasive changes...but, the other flags should be reasonably implementable.

**Comment by dwendt** [comment]
adding my +1, and an additional test case:

`asr.b{*}(bpl, 2)`
```
40 c0 fd 02  sar bpl, 0x2
0f 82 ae f1 05 00   jb xxxxxxx
```

**Comment by anarelion** [comment]
Additional test case

<img width="926" height="308" alt="Image" src="https://github.com/user-attachments/assets/786a6e4c-be9f-4b7d-a5ca-dcf401c88ff2" />

**Comment by anarelion** [comment]
If lets say I wanted to contribute to the project, what would need to be done to fix this?

**Comment by whitequark** [comment]
Would like to ping this on behalf of another BN user. Here is a sample they provided (load as x86, hit P on first byte): 

[splice.zip](https://github.com/user-attachments/files/25754239/splice.zip)

<img width="2014" height="710" alt="Image" src="https://github.com/user-attachments/assets/54c54a9d-128a-4ed2-9b0a-d9a7743a0d11" />

**Comment by robert-yates** [comment]
These missing flag lifts are usually not an issue, however the use of the P flag is currently a good way to obfuscate control flow, for example:

```
   0 @ 10026090  push(ebp)
   1 @ 10026091  ebp = esp {var_4}
   2 @ 10026093  temp1.d = esp {var_4}
   3 @ 10026093  esp = esp - 8
   4 @ 10026093  flag:c = temp1.d u< 8
   5 @ 10026093  flag:p = unimplemented
   6 @ 10026093  flag:z = temp1.d == 8
   7 @ 10026093  flag:s = temp1.d - 8 s< 0
   8 @ 10026093  flag:o = add_overflow(temp1.d, -8)
   9 @ 10026096  push(ebx)
  10 @ 10026097  push(esi)
  11 @ 10026098  push(edi)
  12 @ 10026099  [ebp - 4 {var_8}].d = 0x25a
  13 @ 100260a0  if (not(flag:p)) then 14 @ 0x100260b3 else 15 @ 0x100260a2
  14 @ 100260b3  if (flag:o) then 16 @ 0x100260a3 else 17 @ 0x100260b5
  15 @ 100260a2  if (flag:p) then 14 @ 0x100260b3 else 18 @ 0x100260a4
  16 @ 100260a3  if (flag:s != flag:o) then 30 @ 0x100260b8 else 31 @ 0x100260ab
  ....
```

using a random P flag and constructing a chain of 2 alternating complex paths that eventually converge at the same point, additionally the jumps use bytes of other jumps as targets to prevent patching.
this is annoying because manual UIDF doesnt work, you need to write a custom workflow.

In anycase, implementing the P flag in the arch plugin requires introducing a resulting expression to operate on, this makes things messey, so i assume thats why its not done and considered not worth it :)

**Comment by whitequark** [comment]
The P flag is just popcnt modulo 2, it is not hard to implement.

**Comment by rssor** [comment]
In principle, yes, however the issue with the flag write LLIL callback is that it needs to return an expr index (which gets used as the rhs for `flag:p = X`) -- however, since we lack a top-level popcnt instruction (along with some other increasingly glaring omissions like bswap, lzcnt, etc.) and since intrinsics aren't directly value-producing exprs but top-level instructions themselves, by the time that callback is being fired it's too late to add whole new instructions.

So since architectures can't really use intrinsics inside that callback, they're limited to what can be cobbled together in a single nested expr tree with normal LLIL ops.

**Comment by whitequark** [comment]
Ah, that makes a lot of sense. Should the popcnt, etc, instructions be added?

**Comment by fuzyll** [comment]
Yes, and we've had discussions surrounding adding additional IL instructions internally. But, we don't currently have any specific details or a timeline. After RE//verse this week and our upcoming release, we should be in a better position to look into tackling the underlying requirements for doing something like this more easily. There are more architectures than just x86 being affected, and we do agree it needs addressed.

**Comment by whitequark** [comment]
Thank you for the update, that sounds excellent!

**Comment by plafosse** [comment]
Since popcnt was added here: https://github.com/Vector35/binaryninja-api/commit/d592ed6d we can make some progress here. Thanks for baring with us!

**Comment by whitequark** [comment]
Excited to see it!

**Comment by zznop** [comment]
https://github.com/Vector35/binaryninja-api/pull/8300 for anyone who would like to test
